### ISSUE TITLE
[question] StructureFromKnownPoses: Does the orientation of the frame of the known poses matter?

### ISSUE THREAD

**Issue by Dave-van-der-Meer** [alicevision/meshroom#1128]
**Describe the problem**
I try to reconstruct a scene with known camera poses ordered as a camera rig (stereo camera). The frame in which the known poses are stored is different from the frame that Meshroom uses, so in the end, I managed to rotate the camera orientations on spot, but the entire rig is still rotated by 90 degrees.

How does Meshroom cope with this situation? Is is necessary that the ground of the point cloud is co-planar with the grid inside Meshroom or is there no difference in Meshroom? I ask because the depthMap node causes trouble and crashes. Could the wrong orientation be the source of this?

**Screenshots**
![sim9_rig_with_poses_default_settings](https://user-images.githubusercontent.com/47270949/97677786-1cc32d80-1a93-11eb-860a-27fb4a5288f6.PNG)

**Desktop (please complete the following and other pertinent information):**
 - OS: Windows 10
 - Meshroom version: release version
   - Binary version 2019.2.0

**Additional context**
I am currently working on simulated data, but the rotated frame will also occur when I make real world images as the software for capturing the poses is the same in both cases. If the rotation is the issue, I can try to fix it for both cases.

I have already opened several issues related to my project:

https://github.com/natowi/meshroom_external_plugins/issues/4
https://github.com/alicevision/meshroom/issues/788

**Comment by natowi** [comment]
Is Meshroom able to reconstruct the images of your dataset with the default pipeline (and rig constraint) ?

**Comment by Dave-van-der-Meer** [comment]
Yes, with the default pipeline, the reconstruction works and the orientation of the ground plane matches with the grid. Only when I import the poses with the StructureFromKnownPoses node, there is a rotation. I know that this rotation is based on differences in the coordinate frame between ROS (robot operating system) that gives me then known poses and Meshroom.

My main question is, does the orientation matter? Withing the frame, the orientation of my cameras in relation to their position is correct. It is simple that all poses together are rotated.

Do I need to rotate the frame before importing the poses or can I proceed with what I have now and look for other error sources?

**Comment by natowi** [comment]
The orientation difference is expected and can be fixed by using the SFM transform node with FromKnownCamera. If you use this, is your scene still rotated by 90°?

**Comment by Dave-van-der-Meer** [comment]
With the transform node, I can transform the scene around the Y axis only, is that correct?
Also, it seems that only the cameras rotate, but not the point cloud.

**Comment by natowi** [comment]
Yes, it will only rotate the scene to align with the orientation of the selected image, with the selected image as origin for the coordinate system.

>  Also, it seems that only the cameras rotate, but not the point cloud.

That is not an expected behaviour.

**Comment by Dave-van-der-Meer** [comment]
I made a screenshot of how the rotation takes place:

![sim9_rig_default_rotated](https://user-images.githubusercontent.com/47270949/97800533-381a6e00-1c36-11eb-917d-572b708d27f1.png)

I used the *transformation* option with the following transformation: 0,0,0:90:1
So in theory it should rotate on the same position by an angle of 90 degrees and not scale.
When applying it on the default pipeline, the node does not crash (just the point cloud is not rotating), but when I apply it on the StructureFromKnownPoses node, the SFMTransform node crashes with the following in the log file:

```
Program called with the following parameters:
 * input = "D:/meshroom_work/dataset_9_meshroom_rig_known_poses/MeshroomCache/StructureFromKnownPoses/1108f7d3ebffad91b57065d6cfe7e18a1f5c37e4/sfm.abc"
 * landmarksDescriberTypes = "sift,akaze"
 * method = "transformation"
 * output = "D:/meshroom_work/dataset_9_meshroom_rig_known_poses/MeshroomCache/SfMTransform/482b05cd5ec6d8c517c1c6816af5de95bcef4907/transformedSfM.abc"
 * scale = 1
 * transformation = "0,0,0;90;1"
 * verboseLevel = "trace"

[11:31:15.866619][debug] 0 samples found in this animated xform.
[11:31:17.009282][debug] 0 samples found in this animated xform.
```

**Comment by natowi** [comment]
What is the result with "from known camera" ?

**Comment by Dave-van-der-Meer** [comment]
> In the default pipeline everything should be rotated, not only the cameras.

As you can see in the picture, only the cameras are rotated, not the point cloud. I have no idea why.

> What is the result with "from known camera" ?

I also tried the **auto_from_cameras** which gave me the same result (point cloud not rotated). When using the **from_single_camera** option and provide the name of the image (with or without file extension) it crashes. Also when I enter the cameraID, it crashes.

What is the syntaxt that I need to use when entering the camera? Is it only the name of the image, the name plus file extension? And how do I enter the cameraID? Simply entering the number of the ID?

The log file looks as follows when it crashed with the name of the image:

```
Program called with the following parameters:
 * input = "D:/student_job/dataset_9_meshroom_rig/MeshroomCache/StructureFromMotion/e93c9f954bb00cac7dab5ee3455f95cd02aeb3aa/sfm.abc"
 * landmarksDescriberTypes = "sift,akaze"
 * method = "from_single_camera"
 * output = "D:/student_job/dataset_9_meshroom_rig/MeshroomCache/SfMTransform/50f9cec6076b39bbf3b2b2720a3631f20a1bdd67/transformedSfM.abc"
 * scale = 1
 * transformation = "17"
 * verboseLevel = "trace"

[11:56:52.050593][debug] 0 samples found in this animated xform.
[11:56:52.058622][debug] 0 samples found in this animated xform.
[11:56:53.503656][debug] 0 samples found in this animated xform.
[11:56:53.504643][debug] 0 samples found in this animated xform.
[11:56:53.505705][debug] 0 samples found in this animated xform.
```

**Comment by natowi** [comment]
> When using the from_single_camera option and provide the name of the image (with or without file extension) it crashes. Also when I enter the cameraID, it crashes.

_from_single_camera_ should have worked. (That´s what I wanted to say)

> What is the syntax that I need to use when entering the camera? Is it only the name of the image, the name plus file extension? And how do I enter the cameraID? Simply entering the number of the ID?

You did everything right. There should be no error.

I´m poking in the dark here. My guess is that there is something wrong with your known camera positions.
To narrow it down, _you need systematically test your setup_.

- Can you reproduce the error on a small subset of your dataset?
- Align and compare the sfm from the default and your known dataset.
- Check that you did not mix up something in the process where you apply the correct orientation of your cameras in the knownposes file

- How is your node graph designed? Could there be an error (wrong connected node inputs/outputs)?
- How do you import your known poses?
- Do you use FeatureMatching with "Match from known camera poses" (only in MR2020.x) ? Does it help?

- What cameras do you use?
- How are the cameras and sensors oriented?

**Comment by Dave-van-der-Meer** [comment]
> Can you reproduce the error on a small subset of your dataset?

I did not yet test on a smaller data set.

> Align and compare the sfm from the default and your known dataset.

For testing the transform tool, I was applying it to the default pipeline (SFM done by Meshroom) and there the point cloud doesn't rotate. So I was not even talking about the known poses. I am currently working on rotating the camera frame outside of Meshroom.

> How is your node graph designed? Could there be an error (wrong connected node inputs/outputs)?
> How do you import your known poses?

I import the poses by writing my own cameraInit,sfm file based on the same JSON structure as the original one and following the structure of the cameras.sfm file generated by the SFM node. This new cameraInit.sfm file is then used for a modified pipeline where the SFM node is replaced by the StructureFromKnownPoses node. I then delete all data from the previous nodes except the data from the cameraInit node, there I only replace the cameraInit.sfm file with my own generated file containing the known poses and the camera intriniscs.

> Do you use FeatureMatching with "Match from known camera poses" (only in MR2020.x) ? Does it help?

I work on the MR2019.2 version and I tried to use MR2020.1 but there I was not able to generate the SFM. Not sure why.

> What cameras do you use?

I use a vistual camera in a simulation, so no real camera. But for changing the rotation, this should not matter.
